hera_librarian/async_transfers/local.py

```diff
@@ -0,0 +1,102 @@
+"""
+The local async transfer manager.
+"""
+
+import os
+import shutil
+from pathlib import Path
+from socket import gethostname
+
+from hera_librarian.transfer import TransferStatus
+
+from ..queues import Queue
+from .core import CoreAsyncTransferManager
+
+
+class LocalAsyncTransferManager(CoreAsyncTransferManager):
+    hostnames: list[str]
+
+    transfer_attempted: bool = False
+    transfer_complete: bool = False
+
+    def batch_transfer(self, paths: list[tuple[Path]]):
+        copy_success = True
+
+        for local_path, remote_path in paths:
+            copy_success = copy_success and self.transfer(
+                local_path=local_path, remote_path=remote_path
+            )
+
+        # Set local
+        self.transfer_attempted = True
+        self.transfer_complete = copy_success
+
+        return copy_success
+
+    def transfer(self, local_path: Path, remote_path: Path):
+        """
+        Raises
+        ------
+
+        ValueError
+            If the transfer fails.
+        PermissionError
+            If the permissions cannot be set.
+        """
+
+        # Need to make sure that the the permissions are correctly
+        # set on all files and directories that we copy over.
+        # They should have rw-rw-r-- and rwxrwxr-x permissions.
+
+        # Get the group of the parent.
+        parent_group = remote_path.parent.stat().st_gid
+        # Get this user's uid.
+        uid = os.getuid()
+
+        def set_for_file(file: Path):
+            if file.is_dir():
+                os.chmod(
+                    file,
+                    0o775,
+                )
+            else:
+                os.chmod(file, 0o664)
+
+            os.chown(file, uid=uid, gid=parent_group)
+
+            return
+
+        copy_success = False
+
+        if local_path.is_dir():
+            copy_success = shutil.copytree(local_path, remote_path)
```

Comment:
No, we don't want to do that actually. If the ingested object is a directory, then that directory must be a 'top-level file' and hence have a unique filename. Allowing `dirs_exist_ok` would allow you to have e.g.

```
file_0 = "/my/fake/path/" (directory)
file_1 = "/my/fake/path/new_file.txt" (file)
```

Files cannot live 'inside' each other, as e.g. in this case `file_0` may already have a file called `new_file.txt` in it.